crate_universe/extension.bzl

```
@@ -0,0 +1,224 @@
+            cargo_config = cfg.cargo_config,
+            manifests = {str(k): str(v) for k, v in manifests.items()},
+            manifest_to_path = module_ctx.path,
+        ),
+    )
+
+    splicing_output_dir = tag_path.get_child("splicing-output")
+    cargo_bazel([
+        "splice",
+        "--output-dir",
+        splicing_output_dir,
+        "--config",
+        config_file,
+        "--splicing-manifest",
+        splicing_manifest,
+        "--cargo-lockfile",
+        cargo_lockfile,
+    ])
+
+    # Create a lockfile, since we need to parse it to generate spoke
+    # repos.
+    lockfile_path = tag_path.get_child("lockfile.json")
+    module_ctx.file(lockfile_path, "")
+
+    # cargo-bazel generate takes the lockfile as input, but also writes
+    # to the lockfile. This means that even though nothing changes, the
+    # modified timestamp of the file is updated. Since the lock file is
+    # an input to the rule, this would invalidate the repo rule,
+    # requiring it to be rerun on every invocation.
+    # To solve this, we allow it to touch a copy of the lock file,
+    # rather than the original.
+    cargo_lockfile_copy = tag_path.get_child("copy/Cargo.lock")
+    module_ctx.file(
+        cargo_lockfile_copy,
+        module_ctx.read(cargo_lockfile),
+    )
+    cargo_bazel([
+        "generate",
+        "--cargo-lockfile",
+        cargo_lockfile_copy,
+        "--config",
+        config_file,
+        "--splicing-manifest",
+        splicing_manifest,
+        "--repository-dir",
+        tag_path,
+        "--metadata",
+        splicing_output_dir.get_child("metadata.json"),
+        "--repin",
+        "--lockfile",
+        lockfile_path,
+    ])
+
+    crates_dir = tag_path.get_child(repo_name)
+    _generate_repo(
+        name = repo_name,
+        contents = {
+            "BUILD.bazel": module_ctx.read(crates_dir.get_child("BUILD.bazel")),
+        },
+    )
+
+    contents = json.decode(module_ctx.read(lockfile_path))
+
+    for crate in contents["crates"].values():
+        repo = crate["repository"]
+        if repo == None:
+            continue
+        name = crate["name"]
+        version = crate["version"]
+
+        # "+" isn't valid in a repo name.
+        crate_repo_name = "{repo_name}__{name}-{version}".format(
+            repo_name = repo_name,
+            name = name,
+            version = version.replace("+", "-"),
+        )
+
+        build_file_content = module_ctx.read(crates_dir.get_child("BUILD.%s-%s.bazel" % (name, version)))
+        if "Http" in repo:
+            # Replicates functionality in repo_http.j2.
+            repo = repo["Http"]
+            http_archive(
+                name = crate_repo_name,
+                patch_args = repo.get("patch_args", None),
+                patch_tool = repo.get("patch_tool", None),
+                patches = repo.get("patches", None),
+                remote_patch_strip = 1,
+                sha256 = repo["sha256"],
```

This is an `Option` in the rust code

```suggestion
                sha256 = repo.get("sha256", None),
```